source/opt/trim_capabilities_pass.cpp

```diff
@@ -0,0 +1,229 @@
+    for (auto capability : capabilities) {
+      if (SPV_SUCCESS != grammar.lookupOperand(SPV_OPERAND_TYPE_CAPABILITY, static_cast<uint32_t>(capability), &desc)) {
+        continue;
+      }
+
+      for (uint32_t i = 0; i < desc->numExtensions; ++i) {
+        output.insert(desc->extensions[i]);
+      }
+    }
+
+    return output;
+  }
+} // namespace
+
+TrimCapabilitiesPass::TrimCapabilitiesPass()
+    : supportedCapabilities_(
+          TrimCapabilitiesPass::kSupportedCapabilities.cbegin(),
+          TrimCapabilitiesPass::kSupportedCapabilities.cend()),
+      forbiddenCapabilities_(
+          TrimCapabilitiesPass::kForbiddenCapabilities.cbegin(),
+          TrimCapabilitiesPass::kForbiddenCapabilities.cend()),
+      untouchableCapabilities_(
+          TrimCapabilitiesPass::kUntouchableCapabilities.cbegin(),
+          TrimCapabilitiesPass::kUntouchableCapabilities.cend()),
+      opcodeHandlers_(kOpcodeHandlers.cbegin(), kOpcodeHandlers.cend()) {}
+
+void TrimCapabilitiesPass::addInstructionRequirements(Instruction *instruction, CapabilitySet *capabilities, ExtensionSet *extensions) const {
+  // First case: the opcode is itself gated by a capability.
+  {
+    const spv_opcode_desc_t *desc = {};
+    auto result = context()->grammar().lookupOpcode(instruction->opcode(), &desc);
+    if (result == SPV_SUCCESS) {
+      addSupportedCapabilitiesToSet(desc->numCapabilities, desc->capabilities, capabilities);
+    }
+
+    if (desc->minVersion <= spvVersionForTargetEnv(context()->GetTargetEnv())) {
+      extensions->insert(desc->extensions, desc->extensions + desc->numExtensions);
+    }
+  }
+
+  // Second case: one of the opcode operand is gated by a capability.
+  {
+    const uint32_t operandCount = instruction->NumOperands();
+    for (uint32_t i = 0; i < operandCount; i++) {
+      const auto& operand = instruction->GetOperand(i);
+      // No supported capability relies on a 2+-word operand.
+      if (operand.words.size() != 1) {
+        continue;
+      }
+
+      // No supported capability relies on a literal string operand.
+      if (operand.type == SPV_OPERAND_TYPE_LITERAL_STRING) {
+        continue;
+      }
+
+      const spv_operand_desc_t *desc = {};
+      auto result = context()->grammar().lookupOperand(operand.type, operand.words[0], &desc);
+      if (result != SPV_SUCCESS) {
+        continue;
+      }
+
+      addSupportedCapabilitiesToSet(desc->numCapabilities, desc->capabilities, capabilities);
+    }
+  }
+
+  // Last case: some complex logic needs to be run to determine capabilities.
+  auto[begin, end] = opcodeHandlers_.equal_range(instruction->opcode());
+  for (auto it = begin; it != end; it++) {
+    const OpcodeHandler handler = it->second;
+    auto result = handler(instruction);
+    if (result.has_value()) {
+      capabilities->insert(*result);
+    }
+  }
+}
+
+std::pair<CapabilitySet, ExtensionSet> TrimCapabilitiesPass::DetermineRequiredCapabilitiesAndExtensions() const {
```

Comment:
I'd prefer if we had a struct with meaningful names instead of `std::pair<>`. It is hard to know what you are accessing when some just sees `first` and `second` for the uses.

Reply:
I now see what you did. This is fine.